File: CPP/Lesson_006_binary_tree_preorder_traversal.cpp
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 006 -- Binary Tree Preorder Traversal
// Category   : Trees
// Difficulty : Easy
// Study Plan : Day 3
// =============================================================
//
// QUESTION:
//   Given the root of a binary tree, return the preorder (Root, Left, Right)
//   traversal of its nodes' values.
//
//   Example:
//     Input : root = [1,null,2,3]
//     Output: [1, 2, 3]
// =============================================================

#include <vector>
#include <string>
#include <iostream>
#include <stack>
#include <queue>
#include <algorithm>
using namespace std;

struct TreeNode { int val; TreeNode *left, *right; TreeNode(int v):val(v),left(nullptr),right(nullptr){} };

class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> out; if (!root) return out;
        stack<TreeNode*> st; st.push(root);
        while (!st.empty()) {
            auto* n = st.top(); st.pop(); out.push_back(n->val);
            if (n->right) st.push(n->right);
            if (n->left)  st.push(n->left);
        }
        return out;
    }
};

int main() {
    auto* r = new TreeNode(1); r->right = new TreeNode(2); r->right->left = new TreeNode(3);
    for (int v : Solution().preorderTraversal(r)) cout << v << " ";
    cout << endl;
}

File: Python/Lesson_006_binary_tree_preorder_traversal.py
# =============================================================
# MIT License | @analyticswithharry2026
# GitHub  : https://github.com/analyticswithharry
# YouTube : Analytics with Harry
# =============================================================
# Lesson     : 006 -- Binary Tree Preorder Traversal
# Category   : Trees
# Difficulty : Easy
# Study Plan : Day 3
# =============================================================
#
# QUESTION:
#   Given the root of a binary tree, return the preorder (Root, Left, Right)
#   traversal of its nodes' values.
#
#   Example:
#     Input : root = [1,null,2,3]
#     Output: [1, 2, 3]
# =============================================================

class TreeNode:
    def __init__(self, v=0, l=None, r=None): self.val, self.left, self.right = v, l, r

class Solution:
    def preorderTraversal(self, root):
        if not root: return []
        out, st = [], [root]
        while st:
            n = st.pop(); out.append(n.val)
            if n.right: st.append(n.right)
            if n.left:  st.append(n.left)
        return out

if __name__ == "__main__":
    root = TreeNode(1, None, TreeNode(2, TreeNode(3)))
    print(Solution().preorderTraversal(root))  # [1, 2, 3]
